src/server.py

```diff
@@ -59,6 +60,14 @@ def _play(self):
         self.pipeline.set_state(Gst.State.PLAYING)
         return GLib.SOURCE_REMOVE
 
+    def pause(self):
+        self.pipeline.set_state(Gst.State.PAUSED)
+        self._paused_by_service = True
```

Comment:
I think it would be safer to set `self._paused_by_service = True` this in the `__bus_message_cb` when the state changes from PLAYING to PAUSED in  `__bus_message_cb`. But just be careful with the fact that when the pipeline is also set to the PAUSED state when stopping the sound.


![](http://3.bp.blogspot.com/_kXsaK1-Dz0s/TS_m2b7bh8I/AAAAAAAAAAw/4ZtmE-AMflU/s1600/state_switch.png)

Reply:
Let me see If I understand, but first let me clarify how it is used.. 

`_paused_by_service` is actually used to differentiate between the other ways in which the sound might end up in the paused state ,versus, the explicit action of pausing it via this new `priority` mechanism. So it has to happen only when we pause it this way.

Maybe I am missing something, but If I try to set this in `__bus_message_cb`, how would I make the differentiation?

Reply:
Two variables `_pause_called = True` before setting to True, and set `_paused_by_service = True` **when the element actually goes to the PAUSED state* only if _pause_called == True. That's the idea I have now. I hope to not miss something.

Reply:
Hmm, what `_paused_by_service` tries to do is to determine if that method was called or not, and not really to determine if the player is actually paused or not. So, its basically just `_pause_called` in your example.

Maybe if I rename it to something like `_paused_called` would it be less confusing?

Reply:
ok

Reply:
I think it solved, by decoupling the two things pause/resume from the tracking/flag part. Plus it helped me fix another bug I found.